CI tests and production builds

scjody opened the pull request.

Continuities committed:
Add .circleci/config.yml

Commit:
Rename docker-compose.dev.yml (unchanged so far)

This allows us to use the recommended docker-compose pattern:
* `docker-compose.yml`: base configuration
* `docker-compose.override.yml`: dev mode overrides
* `docker-compose.prod.yml`: prod overrides

https://docs.docker.com/compose/extends/#example-use-case

Commit:
Move dev variables to overrides file

Commit:
Split RUN commands unchanged into multiple lines

Commit:
Remove extra copy of node_modules from image

Moving (rather than copying) node_modules and doing all of this in one
RUN command avoids having two copies of node_modules in the image.

Commit:
Add named dev target to server Dockerfile

Commit:
Include activitypub in api docker build context

This requires changing the context to the toplevel directory of the
project, and allows us to ADD activitypub for production builds.

Commit:
Add prod build and compose section for API

Commit:
Rename Dockerfile

The same Dockerfile is now used for dev and prod.

Commit:
Build API image in CircleCI

Commit:
Build activitypub for production api image

Commit:
Push API image on success

Commit:
Run API tests in CI

Commit:
Add .dockerignore

Continuities commented on api/package.json:
```diff
@@ -8,7 +8,7 @@
     "eslint": "eslint ./src",
     "flow:check": "flow check ./src/",
     "flow:build": "flow-remove-types ./src/ -d ./dist/ --all --pretty",
-    "build": "yarn eslint && yarn flow:check && yarn flow:build",
```
I don't think this is necessary any more. I expect it was due to the top-level import from `dist` in the test file, which is no longer present.

scjody replied:
OK, great. It definitely doesn't seem like the right thing to do in any case.

Continuities commented on api/Dockerfile:
```diff
@@ -0,0 +1,31 @@
+FROM node:latest AS dev
```
Wait, this works? You can define two image build scripts in one Dockerfile? How do you select between them when running dev vs prod environments?

scjody replied:
Yeah, it's relatively new: https://docs.docker.com/develop/develop-images/multistage-build/

By default it builds all targets in the file but you can choose to build a specific target with the `--target` option, which is also supported by docker-compose.

One of the other neat things you can do with multistage builds is to install a bunch of build dependencies and do a build in one image, then copy just the thing that was built into a second image. I plan on doing this for the prod build of the app, so you'll see it in action soon.

Continuities commented on docker/docker-compose.yml:
```diff
@@ -0,0 +1,39 @@
+      - api
+      - app
+
+  app:
+    build: 
+      context: ../app
+      dockerfile: Dockerfile.dev
+    image: den-app
```
Will a prod version of the app be coming in a later PR? We definitely don't want to be running webpack-dev-server in production.

scjody replied:
Yeah, I'm still working on `api` but `app` is on my list.

All the `app` Dockerfile needs to do for production is to build assets that can then be served statically by `nginx`.

scjody replied:
This PR now includes a production version of the app. It builds an nginx image that includes the static files since docker-compose doesn't allow you to build an image and use its files from a different container. (In other words we couldn't build an `app` image and allow `nginx` to serve its contents statically without something weird like wrapping docker-compose in our own build system.)

I'm still working on the nginx config.

Commit:
Move dev-specific build options to dev override

Commit:
Mark activitypub eslintrc as root

This prevents eslint from loading `.eslintrc` files from parent directories
like `app` or `api`, which would otherwise cause linting to behave
differently depending on what subproject it's run from.

This is particularly important because the `.eslintrc` for `app`
loads `eslint-plugin-react`, which isn't a dependency of `activitypub`
(and shouldn't be).

Commit:
Add app production Dockerfile

Commit:
Build app in CI

Commit:
Run babel on activitypub

Commit:
Move http section of nginx config to an include

Commit:
Use nginx to serve static files in prod app
